stumpy/__init__.py

```
@@ -32,6 +35,18 @@
 from .stumped import stumped  # noqa: F401
 from .stumpi import stumpi  # noqa: F401
 
+# Get the default fastmath flags for all njit functions
+# and update the _STUMPY_DEFAULTS dictionary
+
+if not numba.config.DISABLE_JIT:  # pragma: no cover
+    njit_funcs = cache.get_njit_funcs()
+    for module_name, func_name in njit_funcs:
+        module = importlib.import_module(f".{module_name}", package="stumpy")
+        func = getattr(module, func_name)
+        key = module_name + "." + func_name  # e.g., core._mass
+        key = "STUMPY_FASTMATH_" + key.upper()  # e.g., STUMPY_FASTHMATH_CORE._MASS
+        config._STUMPY_DEFAULTS[key] = func.targetoptions.get("fastmath", False)
```

Or should we do `func.targetoptions.get("fastmath", None)`? Then, when it comes to reseting the fastmath flag of the function, we can delete the `fastmath` key from the dictionary `func.targetoptions` if the default value is `None`?

---

Hmm, why would/should it ever be `None`? `func.targetoptions['fastmath']` should either be `True` or a Python set of flags, right? In fact, if there is no `fastmath` key in `func.targetoptions` then I would want Python to error out as this is unexpected behavior.

---

> func.targetoptions['fastmath'] should either be True or a Python set of flags, right?

Only if the value for fastmath flag is passed to the njit decorator. For example,

```
import numba

@numba.njit()
def func():
    return

print(func.targetoptions)
```
shows `{'nopython': True, 'boundscheck': None}`. The default is `False`. 

---

So, I am trying to cover a scenario in which  a user sets fastmath flag for a STUMPY's njit function that has no fastmath by default. And then tries to reset it by doing `cache.reset(...)`

---

More specifically, is there a current STUMPY function that does not have `fastmath` passed to `njit`? And, if so, should it have it? Off the top of my head, I would expect ALL `njit` functions to have a `fastmath` flag by default.

---

I guess this is somewhat related to issue #708 and PR #1025.

> I noticed that your original list did not contain functions like core._get_array_ranges or core._get_ranges or core._merge_topk_PI (and many others). They should all have some sort of fastmath flag

So, there are indeed function without any `fastmath` flags

---

> there are indeed function without any fastmath flags

```
# format: (module_name, func_name)

('core', '_get_array_ranges')
('core', '_get_ranges')
('core', '_merge_topk_PI')
('core', '_merge_topk_ρI')
('core', '_shift_insert_at_index')
```

We can go with different options when there is no fastmath flag.

(1) Add `fastmath=False` for those functions.
(2) Do not add a key for that function in config._STUMPY_DEFAULTS  
(3) Add key and set the value to False
(4) Add key and set the value to None

Both (3) and (4) seem good. I prefer (3) because it reminds us that there was no fastmath flag, and when we reset it, the function will EXACLY become the original STUMPY's function.

---

> Can I address this in a separate, new PR first? and come back here?

Yes please.

---

> So, I am trying to cover a scenario in which a user first sets fastmath flag for a STUMPY's njit function that has no fastmath (i.e. fastmath=False). And then tries to reset it by doing cache.reset(...)

You mean "by doing config._reset()" and not "cache.reset()", right? Because, `cache.py` should have nothing to do with config defaults (i.e., it has no knowledge or authority over the fastmath flags)! `cache.py` functions only (blindly) deal with the cache (e.g., enabling it, clearing it, recompiling it, etc).

---

Oops! I misspoke! I meant "... to reset it by doing `fastmath._reset(...)`".

---

Excellent. That makes a lot more sense. Thank you for clarifying!